### ISSUE TITLE
Video processing

### ISSUE THREAD

**Issue by geekedtv** [zusam/zusam#36]
I uploaded a 1 minute video and it still says "Video is not ready yet." even after waiting over 1 hour. Tried both .mov and .mp4.

**Comment by nrobinaubertin** [comment]
By default, it will convert videos only on "idle hours" that are defined as 1AM-7AM.  
You can change this behavior by creating a `config` file in your `data/` directory and changing the defaults that are listed in the [.env file](https://github.com/zusam/zusam/blob/master/api/.env).

**Comment by geekedtv** [comment]
Ahhhh, very informative. Thank you!

**Comment by Neriderc** [comment]
If I try this, it seems to disable media upload. Even if I don't change the default, if I just add:
`IDLE_HOURS="01-07"`
Into the data/config file, the buttons to upload media disappear.

I'm running the master build. Any tips?

**Comment by nrobinaubertin** [comment]
I'm not able to reproduce your issue

The capabilities of the backend should be reflected at the path `/api/info` (meaning that if your instance is hosted at `https://example.com`, it will be `https://example.com/api/info`). Are the `upload` values set to `true` ?

**Comment by Neriderc** [comment]
Ok this is weird. I rebooted the container and it worked fine. However, when I try to change the quiet hours (with the intention of having no quiet hours and just having videos process when they are uploaded), I tried to set:
`IDLE_HOURS="00-23"`
This makes the media upload buttons disappear. Then changing it to:
`IDLE_HOURS="01-07"`
Doesn't fix it, the buttons are still gone.
But removing the line fixes it.
Do you see the same behaviour? Would "00-23" not be a valid value?

**Comment by nrobinaubertin** [comment]
Whats strange is that IDLE_HOURS is not used to detect the capabilities of the backend (that's why your upload values are true) and you could set it to whatever value and still be able to upload videos (but they would probably not get converted). 

"00-23" is a valid value. It should be something like `/\d+-\d+/` (as a regexp).

The webapp reads the values of `/api/info` to decide wether it should display the media upload buttons. That said, I often have to hard reload the page on a new instance (with shift+r) so that it correctly displays them. Maybe that could fix it on your side ?

**Comment by Neriderc** [comment]
Ok so a bit more experimenting. If I set the value to "00-23", then not just hard refresh the page but actually log out and back in, it appears. But when I then refresh the page... those buttons disappear.

So I've managed to actually have it work, where I've set it to all hours and then log out and back in and the buttons are there. Just so long as I don't refresh that page for any reason, otherwise the buttons will disappear (but come back if I close the post and go back to the group, then click to make a new post again).

**Comment by geekedtv** [comment]
What exactly are the video size limitations before they are put into a schedule for processing? I uploaded a smaller MP4 video and it processed right away. What size is the cutoff before it sends it to the process queue?

**Comment by nrobinaubertin** [comment]
@geekedtv: I arbitrarily decided to not process a video if it's under 10Mo (and a mp4), [here is the code for reference](https://github.com/zusam/zusam/blob/master/api/src/Service/File.php#L57).  

FYI: a processed video will be converted to mp4 (aac audio) at 720p. [Here are the ffmpeg conversion options](https://github.com/zusam/zusam/blob/master/api/src/Command/ConvertVideo.php#L65).  

Those are arbitrary limitations but zusam wasn't meant to be a full-fledge video platform.

**Comment by Neriderc** [comment]
I'm now the day after uploading a 250mb video, and it hasn't been processed. i'm using the docker installation.

Is there a cron I'm supposed to have set up or does the container handle it? The install instructions don't mention cron.

**Comment by nrobinaubertin** [comment]
@Neriderc There's a cron job running in the container that should handle it.  
You can check `/zusam/api/var/log/` for log files inside the container to check if something has gone wrong.  
You can also force the video conversion by doing `docker exec -it zusam php /zusam/api/bin/console zusam:convert:video -vv` (if your container is named zusam). The `-vv` is to set the verbosity level to check what's happening.

**Comment by Neriderc** [comment]
So running the force video conversion gives me:
```
01:05:00 INFO      [zusam] zusam:convert:video
Converting 30346233-8c8d-4c98-b66c-31861edd5dff.mp4
zusam:convert-video 30346233-8c8d-4c98-b66c-31861edd5dff failed.
```

Checking for logs at the location you gave me shows nothing.

**Comment by nrobinaubertin** [comment]
@Neriderc Could you retry to run `docker exec -it zusam php /zusam/api/bin/console zusam:convert:video -vvv` with the update code of the master branch (you'll need to rebuild the container) ?  
I hope that the issue will appear this time

**Comment by Neriderc** [comment]
I have fully cleared everything, downloaded the latest version, rebuilt the container, and run that command. It's now working!
I get the message:
```
01:08:13 INFO      [zusam] zusam:convert:video
01:08:13 NOTICE    [zusam] Converting /zusam/data/files/cf759362-f931-4d8f-b96f-da994777f093.mp4
```
Then after a minute or two it finished, and the video showed in Zusam.

However, setting the idle hours to 00-23 didn't cause this to trigger automatically (the cron seems like this should trigger every minute, but it didn't trigger after several minutes).

Any ideas on how to get that working?

There's still no log, if that helps.

**Comment by nrobinaubertin** [comment]
> However, setting the idle hours to 00-23 didn't cause this to trigger automatically (the cron seems like this should trigger every minute, but it didn't trigger after several minutes).

Although the zusam:cron command is called every minute, it has a minimal period for each type of task. The minimal period for video conversion is [one hour](https://github.com/zusam/zusam/blob/master/api/src/Command/Cron.php#L49).  
I could add a way to change that in the config file.

**Comment by Neriderc** [comment]
Ok makes sense.

I guess if I was going to use this as a sort of private social network for sharing and having conversations with friends and family (which is something I'm looking for a tool to do), people are going to expect that when they upload a video, it's available pretty soon for other people to see. 

I'm guessing I can just change that part in the code and rebuild the container?

Having it in the config file would likely be a good idea, but not sure where it would lie in terms of priority.

I have a few other suggestions that I've come up with while playing with it. Did you want them listed in separate issues or do you have a place you're tracking ideas so that you can individually accept/reject them and prioritise them?

**Comment by nrobinaubertin** [comment]
> I guess if I was going to use this as a sort of private social network for sharing and having conversations with friends and family (which is something I'm looking for a tool to do), people are going to expect that when they upload a video, it's available pretty soon for other people to see.

I understand. I made these choices to accomodate with slow hardware like raspberries and to enable cron-like tasks without having cron (like wordpress, to be able to deploy zusam on a shared hosting environment).  
There maybe a way to do it better. 

> I'm guessing I can just change that part in the code and rebuild the container?

That would work yes

> Having it in the config file would likely be a good idea, but not sure where it would lie in terms of priority.

It's easy to implement, I'll do it when I get the time this week. 

> I have a few other suggestions that I've come up with while playing with it. Did you want them listed in separate issues or do you have a place you're tracking ideas so that you can individually accept/reject them and prioritise them?

I prefer having them listed as separate issues. Thanks !

**Comment by Neriderc** [comment]
> I understand. I made these choices to accomodate with slow hardware like raspberries 

 I'm running on a Raspberry Pi 4 (8GB RAM, and SSD not SD card) and the conversions don't seem too bad, went faster than I expected, and it seems to only be running on one core so more potential there.

> and to enable cron-like tasks without having cron (like wordpress, to be able to deploy zusam on a shared hosting environment).
> There maybe is a way to do it better.

Yeah it makes sense in that context. Do shared hosts often have ffmpeg available though?


> I prefer having them listed as separate issues. Thanks !
Great, I'll add them when I get time over the next few days.

I also just want to say thanks for the help understanding this over the past few days. I've been looking for a forum-like (or social media-like) tool that handles videos well (without having to upload them on youtube or some separate site). Yours is the only one I've found that lets you upload the giant videos that smartphones make these days (like my 40 second 250MB file I've been testing with) and scale it to something web friendly. 

I'd love to contribute but I've afraid my citizen-developer skills would make things take longer than if I didn't try so I might have to stick to providing some ideas and seeing how it shapes up over time.

**Comment by nrobinaubertin** [comment]
> I'm running on a Raspberry Pi 4 (8GB RAM, and SSD not SD card) and the conversions don't seem too bad, went faster than I expected, and it seems to only be running on one core so more potential there.

I will add the option to run on multiple cores also. It's true that raspberries are beasts now !

> Yeah it makes sense in that context. Do shared hosts often have ffmpeg available though?

ffmpeg no, but the image conversion and various other tasks are run via zusam:cron.

> I also just want to say thanks for the help understanding this over the past few days. I've been looking for a forum-like (or social media-like) tool that handles videos well (without having to upload them on youtube or some separate site). Yours is the only one I've found that lets you upload the giant videos that smartphones make these days (like my 40 second 250MB file I've been testing with) and scale it to something web friendly.

> I'd love to contribute but I've afraid my citizen-developer skills would make things take longer than if I didn't try so I might have to stick to providing some ideas and watching how it shapes up over time.

Thanks a lot !  
Knowing that this project can be of any use by others is a big boost for me.  
Having ideas and open issues about bugs are good ways to contribute.

**Comment by Neriderc** [comment]
> I will add the option to run on multiple cores also. It's true that raspberries are beasts now !

That would be great! Please make it not "1 core or all cores", but "1 core, 2 cores, 3 cores", etc. I'd worry using all 4 cores of the Pi would slow everything to a halt during processing, but keen to try 2 or 3 cores and see how it gets on.

**Comment by cmuc24** [comment]
Hi, 
is there a option to turn off Converting Videos ? 
Or how to prevent videos from converting?

**Comment by nrobinaubertin** [comment]
@cmuc24 

You can set `ALLOW_VIDEO_UPLOAD="false"` in the `data/config` file to disable the possibility to upload videos.  

I could also add a parameter to avoid converting videos but that would mean restricting the type of file that can be uploaded (to something that would be configurable but probably mp4/webm only would be wise for compatibilty reasons). 

---

@Neriderc 
> That would be great! Please make it not "1 core or all cores", but "1 core, 2 cores, 3 cores", etc. I'd worry using all 4 cores of the Pi would slow everything to a halt during processing, but keen to try 2 or 3 cores and see how it gets on.

I added the `VIDEO_CONVERSION_THREADS` parameter on the [master branch](https://github.com/zusam/zusam/blob/master/api/.env#L42).

**Comment by cmuc24** [comment]
before turning off all uploads, i ve two other ideas

some "convert" parameter with lower as now or zero impact,
should be given not for all, but for well known videoformat/file.extension 
or
check uploaded videoformat, and if they pass some 'playable' criterias, do not convert.

for users with no converting issues, because almost all clips in .mp4 
and for users with local instances, no bandwidth issues

**Comment by nrobinaubertin** [comment]
> check uploaded videoformat, and if they pass some 'playable' criterias, do not convert.

The actual "playable" criteria is format=mp4 and size<=10Mo.  
I could make those criterias configurable.

**Comment by Neriderc** [comment]
> I added the `VIDEO_CONVERSION_THREADS` parameter on the [master branch](https://github.com/zusam/zusam/blob/master/api/.env#L42).

Awesome! I also noticed the config options for changing the conversion delays. I set the threads to 3, set the cron video setting to 60 seconds, and tested by uploading my 250MB video. It triggered off after about a minute, with ffmpg clearly using heaps of CPU. This is on my Raspberry Pi 4. I then tried using Zusam while that was running, and it was still very usable. I thought it might get slow with too much load by 3 out of 4 cores on conversion seems to leave enough for other processes.

It did take a little longer than I expected, using 3 cores didn't cut it to 1/3 of the time but that would probably be an unrealistic expectation. It was about 15 mins to do that video, which I'm pretty happy with.
